server/study_server.js

```diff
@@ -153,56 +153,76 @@ module.exports = {
     },
     submitResult: function (req, res, next) {
         var clean_id = sanitize(req.body.id);
+        var max_retries = 3;
 
-         Study.findOne({_id: clean_id}, function (err, study) {
-            if (err) {
-                res.status(504);
-                logger.error("study_server.js: Error in submit result:", err);
-                res.end(err);
-            } else {
-                //if the study is being previewed, don't record the response
-                if (req.body.resid === "preview"){
-                    res.redirect('/studies');
+        //if the study is being previewed, don't record the response
+        if (req.body.resid === "preview") {
+            res.redirect('/studies');
+            res.end();
+        }
+
+        // function to save complete response to study and remove from incomplete responses
+        function saveStudy(study, responseID, retries){
+            var respIdx = study.incompleteResponses.indexOf(responseID);
+            study.completeResponses.push(study.incompleteResponses[respIdx]);
+            study.incompleteResponses.splice(respIdx,1);
```

Comment:
There is definitely a race condition going on here.

We have potentially multiple requests that can be processed simultaneously which would cause an inconsistent study states if we don't use a proper locking mechanism.

1) If two concurrent requests submit for the same study, both might find the `respIdx` in `incompleteResponses` and try to move it to `completeResponses`. This will cause one of the request to fail to update due to it.
2) Removing `respIdx` from `incompleteResponses` that has already been removed by another request is another issue. Using splice based on an index found earlier is more dangerous than filtering the array directly, because things can happen between retrieval and update... that must be done atomically!

Reply:
I have not spent enough time thinking that in details, but please double check if the operation is safe or not.

Reply:
https://github.com/ScottLogic/kort/blob/3282d6227346af2a7512ba9ff1863757cf0cae4f/server/study_server.js#L157

I feel that's where we need to use an atomic operation such as `findOneAndUpdate()` which will lock the DB while we do our processing/filtering

Reply:
Are the responseId unique per request?

Reply:
I was under the impression that you should use save() where possible when updating a document (link: https://mongoosejs.com/docs/tutorials/findoneandupdate.html) 

ResponseIDs are unique throughout the collection of responses.